// kinto-http/lib/cjs-es5/batch.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Exports batch responses as a result object.
 *
 * @private
 * @param  {Array} responses The batch subrequest responses.
 * @param  {Array} requests  The initial issued requests.
 * @return {Object}
 */
function aggregate(responses, requests) {
    if (responses === void 0) { responses = []; }
    if (requests === void 0) { requests = []; }
    if (responses.length !== requests.length) {
        throw new Error("Responses length should match requests one.");
    }
    var results = {
        errors: [],
        published: [],
        conflicts: [],
        skipped: [],
    };
    return responses.reduce(function (acc, response, index) {
        var status = response.status;
        var request = requests[index];
        if (status >= 200 && status < 400) {
            acc.published.push(response.body);
        }
        else if (status === 404) {
            // Extract the id manually from request path while waiting for Kinto/kinto#818
            var regex = /(buckets|groups|collections|records)\/([^/]+)$/;
            var extracts = request.path.match(regex);
            var id = extracts && extracts.length === 3 ? extracts[2] : undefined;
            acc.skipped.push({
                id: id,
                path: request.path,
                error: response.body,
            });
        }
        else if (status === 412) {
            acc.conflicts.push({
                // XXX: specifying the type is probably superfluous
                type: "outgoing",
                local: request.body,
                remote: (response.body.details && response.body.details.existing) || null,
            });
        }
        else {
            acc.errors.push({
                path: request.path,
                sent: request,
                error: response.body,
            });
        }
        return acc;
    }, results);
}
exports.aggregate = aggregate;
