// kinto-http/lib/cjs-es5/endpoint.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Endpoints templates.
 * @type {Object}
 */
var ENDPOINTS = {
    root: function () { return "/"; },
    batch: function () { return "/batch"; },
    permissions: function () { return "/permissions"; },
    bucket: function (bucket) { return "/buckets" + (bucket ? "/" + bucket : ""); },
    history: function (bucket) { return ENDPOINTS.bucket(bucket) + "/history"; },
    collection: function (bucket, coll) {
        return ENDPOINTS.bucket(bucket) + "/collections" + (coll ? "/" + coll : "");
    },
    group: function (bucket, group) {
        return ENDPOINTS.bucket(bucket) + "/groups" + (group ? "/" + group : "");
    },
    record: function (bucket, coll, id) {
        return ENDPOINTS.collection(bucket, coll) + "/records" + (id ? "/" + id : "");
    },
    attachment: function (bucket, coll, id) {
        return ENDPOINTS.record(bucket, coll, id) + "/attachment";
    },
};
exports.default = ENDPOINTS;
